src/dma/mod.rs

```diff
@@ -0,0 +1,1171 @@
+            buf: Some(memory),
+            double_buf,
+            transfer_length: n_transfers,
+        };
+        transfer.apply_config(config);
+
+        Ok(transfer)
+    }
+
+    /// Starts the transfer, the closure will be executed right after enabling the stream.
+    pub fn start<F>(&mut self, f: F)
+    where
+        F: FnOnce(&mut PERIPHERAL),
+    {
+        // "Preceding reads and writes cannot be moved past subsequent writes"
+        compiler_fence(Ordering::Release);
+
+        unsafe {
+            self.stream.enable();
+        }
+        f(&mut self.peripheral);
+    }
+
+    /// Pauses the dma stream, the closure will be executed right before disabling the stream.
+    pub fn pause<F>(&mut self, f: F)
+    where
+        F: FnOnce(&mut PERIPHERAL),
+    {
+        f(&mut self.peripheral);
+        self.stream.disable()
+    }
+
+    /// Changes the buffer and restarts or continues a double buffer transfer. This must be called
+    /// immediately after a transfer complete event. Returns the old buffer together with its
+    /// `CurrentBuffer`, if an error occurs, this methods will return the new buffer with the error.
+    pub fn next_transfer(
+        &mut self,
+        mut new_buf: BUF,
+    ) -> Result<(BUF, CurrentBuffer), DMAError<BUF>> {
+        if !STREAM::get_transfer_complete_flag() {
+            return Err(DMAError::NotReady(new_buf));
+        }
+        self.stream.clear_transfer_complete_interrupt();
+
+        if self.double_buf.is_some() && DIR::direction() != DmaDirection::MemoryToMemory {
+            // NOTE(unsafe) We now own this buffer and we won't call any &mut methods on `new` until
+            // the end of the DMA transfer
+            let (new_buf_ptr, new_buf_len) = unsafe { new_buf.write_buffer() };
+
+            // We can't change the transfer length while double buffering
+            if new_buf_len < usize::from(self.transfer_length) {
+                return Err(DMAError::SmallBuffer(new_buf));
+            }
+
+            if STREAM::current_buffer() == CurrentBuffer::DoubleBuffer {
+                self.stream.set_memory_address(new_buf_ptr as u32);
+                let old_buf = self.buf.replace(new_buf);
+
+                // "Subsequent reads and writes cannot be moved ahead of preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                // We always have a buffer, so unwrap can't fail
+                return Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer));
+            } else {
+                self.stream
+                    .set_memory_double_buffer_address(new_buf_ptr as u32);
+                let old_buf = self.double_buf.replace(new_buf);
+
+                // "Subsequent reads and writes cannot be moved ahead of preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                // double buffering, unwrap can never fail
+                return Ok((old_buf.unwrap(), CurrentBuffer::DoubleBuffer));
+            }
```

Comment:
In that case the `stream.set_memory_address(...)` would fail, but that is still a problem, since the user will be given the buffer to where the DMA is writing, the check is right before changing the memory in this case, so like a few clock cycles, this seems to be almost impossible to cause problems, also checking again after can cause some false positives, and in the case of a false positive we would have the same problem, since we would give the `new_buf` back with the error, but that buffer would be the one the DMA would be writing to, so not sure what we can do here... panic ?

Reply:
We found a solution in Matrix: Since the MAR registers are write-protected if the respective buffer is currently used by DMA we can simply check if they contain the address of the new buffer after writing it. And if not, raise/return an Overrun error.